### PR TITLE
8172366: Support SHA-3 based signatures

### PR DESCRIPTION
**Pull request by valeriepeng**
Could someone please help review this RFE?

Enhance default JDK providers except SunPKCS11 with signatures using SHA-3 family of digests. SunPKCS11 provider will be updated separately (JDK-8242332).

This changes covers SUN, SunRsaSign, and SunEC providers. Changes are straightforward, just add SHA-3 digests to various signature algorithms. 

Please review the corresponding CSR as well. It's at: https://bugs.openjdk.java.net/browse/JDK-8252260

Thanks!
Valerie

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8172366](https://bugs.openjdk.java.net/browse/JDK-8172366): Support SHA-3 based signatures

### Reviewers
 * [Xue-Lei Andrew Fan](https://openjdk.java.net/census#xuelei) (@XueleiFan - **Reviewer**) ⚠️ Review applies to 62fb5a39eb0895e68417c2c8a87df5b9c416f918

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/102/head:pull/102`
`$ git checkout pull/102`

### CODE REVIEW

**Comment by XueleiFan** [comment]
Do you want to list the "inP1363Format" algorithms in the "Java Security Standard Algorithm Names" documentation in this CSR?

File: src/java.base/share/classes/sun/security/provider/DSA.java [hunk 1]
```diff
@@ -49,9 +49,11 @@
...
  * This file contains both the signature implementation for the
  * commonly used SHA1withDSA (DSS), SHA224withDSA, SHA256withDSA,
- * as well as RawDSA, used by TLS among others. RawDSA expects
- * the 20 byte SHA-1 digest as input via update rather than the
- * original data like other signature implementations.
+ * SHA384withDSA, SHA512withDSA, SHA3-224withDSA, SHA3-256withDSA,
+ * SHA3-384withDSA, SHA3-512withDSA, as well as RawDSA, used by
+ * TLS among others. RawDSA expects the 20 byte SHA-1 digest as
+ * input via update rather than the original data like other signature
+ * implementations.
```
Comment (XueleiFan) [inline src/java.base/share/classes/sun/security/provider/DSA.java:56]: As you were already there, do you want to mention the IEEE P1363 format in the class description?  Maybe in the 1st paragraph in the class spec.

**Comment by valeriepeng** [comment]
> Do you want to list the "inP1363Format" SHA3 algorithms in the "Java Security Standard Algorithm Names" documentation in this CSR?

Yes, I will do that. Thanks~

File: src/java.base/share/classes/sun/security/provider/DSA.java (hunk 1 above)
Comment (valeriepeng) [reply src/java.base/share/classes/sun/security/provider/DSA.java:56]: Ok~

File: src/java.base/share/classes/sun/security/provider/DSA.java [hunk 2]
```diff
@@ -504,6 +506,50 @@ public String toString() {
         return printable;
     }
 
+    public static final class SHA3_224withDSA extends DSA {
+        public SHA3_224withDSA() throws NoSuchAlgorithmException {
+            super(MessageDigest.getInstance("SHA3-224"));
+        }
+    }
+    public static final class SHA3_224withDSAinP1363Format extends DSA {
```
[the remark this answers is not in the document]
Comment (valeriepeng) [reply src/java.base/share/classes/sun/security/provider/DSA.java:514]: Ok.

**Review by XueleiFan — APPROVED** [review]
Looks good to me,

**Review by XueleiFan — APPROVED** [review]
Looks good.

File: src/java.base/share/classes/sun/security/rsa/RSAPSSSignature.java [hunk 3]
```diff
@@ -272,7 +276,7 @@ private PSSParameterSpec validateSigParams(AlgorithmParameterSpec p)
         // check key length again
         if (key != null) {
             try {
-                int hLen = DIGEST_LENGTHS.get(digestAlgo);
+                int hLen = DIGEST_LENGTHS.get(KnownOIDs.findMatch(digestAlgo));
```
[the remark this answers is not in the document]
Comment (valeriepeng) [reply src/java.base/share/classes/sun/security/rsa/RSAPSSSignature.java]: Sure, will update.

**Review by seanjmullan — COMMENTED** [review]
The new constants in MGF1ParameterSpec.java should have "@since 16".

**Comment by valeriepeng** [comment]
/summary Enhance default JDK providers including SUN, SunRsaSign, and SunEC, with signatures using SHA-3 family of digests.

**Comment by valeriepeng** [comment]
> 
> 
> The new constants in MGF1ParameterSpec.java should have "@SInCE 16".

Yes, made the change in last commit.

**Pull request closed without merging**